src/levanter/models/backpack.py

```diff
@@ -114,8 +115,9 @@ def init(config: Gpt2Config, *, key) -> "WeightsOnlyAttention":
         k_c, _ = jrandom.split(key, 2)
         c_attn = hnn.Linear.init(In=Embed, Out=(Qk, config.Senses, config.SenseHeadDim), key=k_c, use_bias=use_bias)
         dropout = hnn.Dropout(config.attn_pdrop)
+        ln = hnn.LayerNorm.init(config.Senses, eps=config.layer_norm_epsilon)
```

Comment:
Do we need this LayerNorm to be on the entire output axises, `(Qk, config.Senses, config.SenseHeadDim)`?

Reply:
also Embed for the same reason i think

Reply:
If I use Embed, I would get an error:
```
FAILED tests/test_backpack.py::test_backpack_predict - ValueError: axis (Axis(name='embed', size=768),) is not in (Axis(name='position', size=512), Axis(name='senses', size=16), Axis(name='key_position', size=512))
```

So based on the principle that we want to apply LayerNorm per senses, it should be normalized along the axises of Position and KeyPosition:
```
ln = hnn.LayerNorm.init((config.Pos, config.KeyPos), eps=config.layer_norm_epsilon)
```

Reply:
i don't think this is right either (in particular, this lets information leak from the future into the past). I now think putting a layer norm in after the final attention isn't going to work. I think we should aggressively layer norm the sense vectors as they come out of the sense network. The issue is how to make them have large enough magnitude, but i think in high dimension it maybe not doesn't matter.

Let's defer to JohnH.

Reply:
Pasting from slack for reference

Layer Norm after the Q and K in the final attention seems fine to me. We can also Layer Norm the sense vector outputs; shouldn't the bias and gain help with the magnitude?

Reply:
If I understand correctly, I think this is what we need for LN on Q and K in the final attention: https://github.com/stanford-crfm/levanter/pull/172/commits/03e0fcfad33b60b2a154041c277c6114be2d5fa3

Reply:
Seems right.